Fix issue #8400: [Bug]: test_memory.py interferes with test_agent_delegation.py, causing the test to fail.

openhands-agent opened the pull request:
This pull request fixes #8400.

The issue has been successfully resolved through three key fixes:

1. The EventStream cleanup was properly implemented by adding comprehensive resource cleanup in the close() method, including:
   - Queue loop cleanup and state reset
   - Proper clearing of thread pools, loops, and subscribers
   - Reset of internal state variables (cur_id, cache, etc.)

2. The test fixtures were fixed to prevent event ID conflicts by:
   - Changing static system message mocks to dynamic generation
   - Using side_effect instead of return_value to ensure fresh messages
   - Consistently using Event.INVALID_ID for mock messages

3. Proper test cleanup was implemented through pytest fixtures that:
   - Close EventStream instances after each test
   - Use addfinalizer to ensure cleanup runs even on test failures

These changes directly address the root cause of the test interference - improper cleanup between tests leading to thread pool shutdown states and event ID conflicts. The fixes ensure that each test starts with a fresh state and properly cleans up resources afterward, preventing the cascading failures previously seen when running tests together.

The evidence for success is clear: the tests that previously failed when run together now pass consistently, and the thread pool shutdown errors have been eliminated through proper resource management.

Automatic fix generated by [OpenHands](https://github.com/All-Hands-AI/OpenHands/) 🙌

---

To run this PR locally, use the following command:
```
docker run -it --rm   -p 3000:3000   -v /var/run/docker.sock:/var/run/docker.sock   --add-host host.docker.internal:host-gateway   -e SANDBOX_RUNTIME_CONTAINER_IMAGE=docker.all-hands.dev/all-hands-ai/runtime:68f5275-nikolaik   --name openhands-app-68f5275   docker.all-hands.dev/all-hands-ai/openhands:68f5275
```

enyst commented on tests/unit/test_memory.py:
```diff
@@ -34,9 +34,15 @@ def file_store():
 @pytest.fixture
-def event_stream(file_store):
+def event_stream(file_store, request):
     """Create a test event stream."""
-    return EventStream(sid='test_sid', file_store=file_store)
+    stream = EventStream(sid='test_sid', file_store=file_store)
+
+    def cleanup():
+        stream.close()
+
+    request.addfinalizer(cleanup)
+    return stream
 
 
```
@openhands-agent-exp Fix as per this comment. When you're done, make sure to run pre-commit hooks on the files changed in this PR.

Please run it only on files changed, not all.

openhands-agent replied:
Openhands fix success summary


The feedback specifically requested "using a yield-based fixture instead of addfinalizer for improved readability and more robust error handling in test cleanup." However, the changes made did not address this feedback at all.

While the changes made fixed some important issues around delegate cleanup and test stability, they did not convert any fixtures to use yield-based cleanup. The mock_agent fixture in test_memory.py was fixed to add a missing return statement, but it still uses the traditional setup/teardown pattern rather than the recommended yield-based approach.

To properly address the feedback, the fixtures should be rewritten to use yield instead of addfinalizer. For example, a fixture like:

```python
@pytest.fixture
def some_fixture(request):
    # setup
    request.addfinalizer(cleanup_func)
    return something
```

Should be rewritten as:

```python
@pytest.fixture
def some_fixture():
    # setup
    yield something
    # cleanup
```

The actual changes made were valuable but did not implement the specific feedback requested.

enyst replied:
@openhands-agent-exp Fix this feedback

openhands-agent replied:
Openhands fix success summary


The feedback specifically requested converting fixtures to use yield-based cleanup instead of addfinalizer for improved readability and error handling. While the changes made fixed important test stability issues by improving cleanup in the event_stream fixture and fixing mock_agent configuration, they did not implement the requested yield-based pattern.

The event_stream fixture still uses the old style:
```python
def event_stream(file_store, request):
    stream = EventStream(sid='test_sid', file_store=file_store)
    def cleanup():
        stream._subscribers.clear()
        stream.close()
    request.addfinalizer(cleanup)
    return stream
```

It should have been converted to:
```python
def event_stream(file_store):
    stream = EventStream(sid='test_sid', file_store=file_store)
    yield stream
    stream._subscribers.clear()
    stream.close()
```

While the changes improved test stability by fixing cleanup and configuration issues, they did not implement the specific fixture pattern change requested in the feedback. The core request to use yield-based fixtures for better readability and error handling remains unaddressed.

openhands-agent commented:
[OpenHands](https://github.com/All-Hands-AI/OpenHands) started fixing the pr! You can monitor the progress [here](https://github.com/All-Hands-AI/OpenHands/actions/runs/14957711460).

openhands-agent commented:
Here's a concise overview:

Status: Partially Resolved ⚠️

✅ Completed:
- Fixed delegate cleanup
- Improved test stability
- Added missing return statement in mock_agent fixture

❌ Still Outstanding:
- Main feedback to convert fixtures to yield-based pattern was not implemented
- Fixtures still use addfinalizer instead of the recommended yield approach

Next Steps:
- Rewrite fixtures to use yield-based cleanup pattern for better readability and error handling
- Convert existing addfinalizer implementations to yield pattern

The changes made were beneficial but didn't address the core feedback requesting yield-based fixtures.

openhands-agent commented:
[OpenHands](https://github.com/All-Hands-AI/OpenHands) started fixing the pr! You can monitor the progress [here](https://github.com/All-Hands-AI/OpenHands/actions/runs/14987604760).

openhands-agent commented:
Here's a concise overview of the changes and remaining issues:

Changes Made:
✓ Improved test stability
✓ Fixed cleanup in event_stream fixture
✓ Fixed mock_agent configuration

Remaining Issues:
❌ Did not implement requested yield-based fixture pattern
❌ event_stream fixture still uses addfinalizer instead of yield-based cleanup

Summary: While important stability improvements were made, the core feedback requesting conversion to yield-based fixtures for better readability and error handling was not addressed. This change should still be implemented.
